Add support for redirected VTL0 interrupts

ricardon opened the pull request:
Hi,

This pull request implements https://github.com/microsoft/openvmm/issues/406. The issue describes the problem statement and proposed solution. 

This code enables the changes made here: https://github.com/microsoft/openvmm/pull/384.

Thank you for your consideration.

ricardon committed:
hv: mshv_vtl: Introduce the MAP_REDIRECTED_DEVICE_INTERRUPT IOCTL

The proxy interrupts that the VTL2 kernel relays to the VTL0 guest are
handled via the single interrupt vector of the VMBus synthetic interrupt
controller.

On TDX guests in particular, Hyper-V does not use IOMMU posted interrupts
for these interrupts. This results in a TDEXIT for every interrupt and
consequent performance degradation.

To recoup performance, newer versions of Hyper-V feature support the
posting of VTL0 interrupts directly to the VTL2 guest. In such setup, the
VTL2 guest still relays interrupts to the VTL0 guest, but uses a dedicated
VTL2 interrupt vector for each VTL0 interrupt. The dedicated vector must be
reserved only in one CPU that the user space specifies. Hyper-V shall post
the interrupt only to the designated VTL2 vCPU. The VTL2 kernel must not
migrate the dedicated vector to another VTL2 vCPU.

User space in the VTL2 guest is responsible of requesting such redirection.
Implement the new MAP_REDIRECTED_DEVICE_INTERRUPT IOCTL. User space
provides the vector number of the VTL0 guest that wants the VTL2 kernel to
redirect as well as the APIC ID of the VTL2 vCPU on which the interrupt
redirection will take place. On success, return in the payload of the IOCTL
the VTL2 CPU vector number reserved for this purpose. This is the vector
number that Hyper-V must use when setting up the posted interrupt.

Reuse the same IOCTL to also request the removal of the redirection when
needed. When removing a redirection user space must provide the number of
the VTL2 interrupt vector number that the kernel provided when the
redirection was established.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * Modified the payload of the IOCTL to also include the APIC ID of the
   VTL2 CPU on which the interrupt redirection will take place.
 * Removed weak stubs that are no longer needed.

Changes since v1:
 * Assigned 0x39 to MAP_REDIRECTED_DEVICE_INTERRUPT as 0x38 was already
   taken. (Roman)
 * Fixed error handling for copy_to_user(). (Roman)
 * Moved comment about unmapping above the conditional statement. (Roman)

ricardon committed:
hv: mshv_vtl: Add a helper function to assert single proxy interrupts

The statically assigned interrupt vector of the VMBus synthetic interrupt
is used to also assert proxy interrupts. These can be asserted in group or
individually. Proxy interrupts can also be asserted individually if a
redirection is in place to use a dedicated Linux IRQ.

Add a helper function to assert individual proxy interrupts.

No functional change intended.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * None

Changes since v1:
 * Rebased on the previous changeset that optimizes the calculation of
   the proxy interrupt bit.

ricardon committed:
hv: mshv_vtl: Create a thin irq_domain for redirected proxy interrupts

Add a thin irq_domain for redirected interrupts. The kernel IRQ
infrastructure will use this domain to manage the resources and
configuration of Linux IRQs to meet the requirements of redirected
proxy interrupts.

The redirected_intr_domain irq_domain defines the .alloc() operation to
request the allocation of IRQs in the IRQ hierarchy all the way up to
x86_vector_domain. Use dummy_irq_chip as the irq_chip for the domain since
no other hardware configuration is needed. Importantly, this irq_chip does
not define the .irq_set_affinity() operation. Redirected proxy interrupts
do not allow changes in affinity.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * Introduced this patch.

Changes since v1:
 * N/A

ricardon commented:
@microsoft-github-policy-service agree [company="Intel Corporation"]


From: microsoft-github-policy-service[bot] ***@***.***>
Sent: Thursday, May 22, 2025 12:47 PM
To: microsoft/OHCL-Linux-Kernel ***@***.***>
Cc: Neri, Ricardo ***@***.***>; Mention ***@***.***>
Subject: Re: [microsoft/OHCL-Linux-Kernel] Add support for redirected VTL0 interrupts (PR #80)

[https://avatars.githubusercontent.com/in/95686?s=20&v=4]microsoft-github-policy-service[bot] left a comment (microsoft/OHCL-Linux-Kernel#80)<https://github.com/microsoft/OHCL-Linux-Kernel/pull/80#issuecomment-2901909239>

@ricardon<https://github.com/ricardon> please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.

@microsoft-github-policy-service agree [company="{your company}"]

Options:

  *   (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.

@microsoft-github-policy-service agree

  *   (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.

@microsoft-github-policy-service agree company="Microsoft"
Contributor License Agreement
Contribution License Agreement

This Contribution License Agreement (“Agreement”) is agreed to by the party signing below (“You”),
and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your
contributions to Microsoft open source projects. This Agreement is effective as of the latest signature
date below.

  1.  Definitions.
“Code” means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to Microsoft under this Agreement.
“Project” means any of the projects owned or managed by Microsoft and offered under a license
approved by the Open Source Initiative (www.opensource.org<http://www.opensource.org>).
“Submit” is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”
“Submission” means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.
  2.  Your Submission. You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.
  3.  Originality of Work. You represent that each of Your Submissions is entirely Your original work.
Should You wish to Submit materials that are not Your original work, You may Submit them separately
to the Project if You (a) retain all copyright and license information that was in the materials as You
received them, (b) in the description accompanying Your Submission, include the phrase “Submission
containing materials of a third party:” followed by the names of the third party and any licenses or other
restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
guidelines concerning Submissions.
  4.  Your Employer. References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.
  5.  Licenses.

  *   Copyright License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
parties.
  *   Patent License. You grant Microsoft, and those who receive the Submission directly or
indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.
  *   Other Rights Reserved. Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

  1.  Representations and Warranties. You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
  2.  Notice to Microsoft. You agree to notify Microsoft in writing of any facts or circumstances of which
You later become aware that would make Your representations in this Agreement inaccurate in any
respect.
  3.  Information about Submissions. You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.
  4.  Governing Law/Jurisdiction. This Agreement is governed by the laws of the State of Washington, and
the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
defenses of lack of personal jurisdiction and forum non-conveniens.
  5.  Entire Agreement/Assignment. This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

—
Reply to this email directly, view it on GitHub<https://github.com/microsoft/OHCL-Linux-Kernel/pull/80#issuecomment-2901909239>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAEV22GZBESN5S6QTJOGZMT27X5ORAVCNFSM6AAAAAB5WSPRFCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBRHEYDSMRTHE>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>

ricardon commented:
@microsoft-github-policy-service agree [company="Intel Corporation"]

ricardon commented:
@microsoft-github-policy-service agree company="Intel Corporation"

romank-msft commented on include/uapi/linux/mshv.h:
```diff
@@ -359,6 +363,8 @@ struct mshv_kick_cpus {
 /* For x86-64 TDX only */
 #define MSHV_VTL_TDCALL _IOWR(MSHV_IOCTL, 0x32, struct mshv_tdcall)
 #define MSHV_VTL_READ_VMX_CR4_FIXED1 _IOR(MSHV_IOCTL, 0x33, __u64)
+#define MSHV_VTL_MAP_REDIRECTED_DEVICE_INTERRUPT _IOWR(MSHV_IOCTL, 0x38, \
```
that was taken recently by the KICK_CPUS ioctl. If you've rebased, please change to a free code.

ricardon replied:
Sure. I didn't realize. I will update it.

romank-msft commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1930,6 +1943,35 @@ static inline long mshv_vtl_ioctl_kick_cpu(void __user *user_arg)
+		 * vector is mapped.
+		 */
+		intr_data.vector = ret;
+		ret = copy_to_user(user_arg, &intr_data, sizeof(intr_data));
```
This might fail returning the number of bytes that weren't copied. Is that good to return from this ioctl?

ricardon replied:
Right. This does not look correct. I'll fix the error handling.

romank-msft commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -1930,6 +1943,35 @@ static inline long mshv_vtl_ioctl_kick_cpu(void __user *user_arg)
+		return (long)-EFAULT;
+
+	if (!intr_data.create_mapping)
+		/* User space provides the hardware vector to unmap. */
```
maybe move the comment above the `if` statement?

ricardon replied:
Sure. Will do.

romank-msft commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -430,13 +430,30 @@ static int mshv_vtl_get_vsm_regs(void)
 
+static void do_assert_single_proxy_intr(const u32 vector, struct mshv_vtl_run *run)
+{
+	const u32 bank = vector / 32;
```
the compiler should optimize this, but maybe `>>5` would be good here?

ricardon replied:
Sure. This is a good improvement. I will update it.

romank-msft commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -430,13 +430,30 @@ static int mshv_vtl_get_vsm_regs(void)
+static void do_assert_single_proxy_intr(const u32 vector, struct mshv_vtl_run *run)
+{
+	const u32 bank = vector / 32;
+	const u32 masked_irr = BIT(vector % 32) & ~READ_ONCE(run->proxy_irr_blocked[bank]);
```
`& 0x1f` instead of the modulo? Thinkling if the compiler doesn't optimize, that's the hot path

ricardon replied:
Sure. Good suggestion. I will add it.

romank-msft commented on arch/x86/kernel/cpu/mshyperv.c:
```diff
@@ -708,6 +766,71 @@ static bool hv_sev_es_hcall_finish(struct ghcb *ghcb, struct pt_regs *regs)
+		return -ENOENT;
+
+	raw_spin_lock_irqsave(&redirected_proxy_intr_lock, flags);
+	redirected_proxy_intr[vec_idx] = REDIRECTED_VECTOR_UNDEF;
```
would `WRITE_ONCE` be enough here instead of the spin lock?

romank-msft replied:
I do see that the map function above does quite a lot but do these map and unmap operations run concurrently?

romank-msft replied:
or other CPUs might be accessing that?

ricardon replied:
Yes. The objective of the spin lock is to serialize access to redirected_proxy_intr[]. Since we access it from interrupt context. I am looking into using a read_lock instead, since when in interrupt we only read the array.

ricardon committed:
hv: mshv_vtl: Optimize computation of the proxy interrupt bit

The computation of the proxy interrupt number needs to be performed each
time is asserted. Optimize the computation using bit shifts and masks.

While here, fix a minor issue on a comment.

Suggested-by: Roman Kisel <romank@linux.microsoft.com>
Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * None

Changes since v1:
 * Introduced this patch.

ricardon commented:
I updated the patches to incorporate feedback from @romank-msft and use a seqcount to serialize access to the redirected proxy interrupts data structures.

romank-msft commented:
LGTM. Someone more familiar with the intent might want to take a look.

ricardon committed:
mshv_vtl/tdx: Add a helper function to get CPU number from APIC ID

mshv_tdx_set_cpumask_from_apicid() does two things: it traverses the APICID
hashtable and sets a CPU mask. Traversing the hastable can be carved out
into a helper function. This makes the code more modular and other
functions can benefit from such helper. Subsequent patches will make use
of it.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * Introduced this patch.

Changes since v1:
 * N/A

ricardon committed:
hv: mshv_vtl: Redirect a proxy interrupt

Implement the setup portion of the MAP_REDIRECTED_DEVICE_INTERRUPT
IOCTL. To redirect a proxy interrupt, allocate and request a Linux IRQ.

Set the CPU affinity of the IRQ to that user space specified. Also, specify
to request_irq() that the interrupt should be excluded from interrupt
balancing as such action would change the CPU affinity of the interrupt.
The contract with user space restricts the kernel from changing it.

Upon getting an interrupt, the handler will assert the proxy interrupt
vector that user space specified.

Remember which hardware vector (i.e., the CPU vector number) that was
assigned to the IRQ. This information is needed to tear down the
redirection of the proxy interrupt (it will be implemented in a subsequent
changset).

Keep a list of the redirected interrupts. Use it to avoid redirecting
the same proxy interrupt more than once.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * Introduced this patch.

Changes since v1:
 * N/A

ricardon committed:
hv: mshv_vtl: Handle unmapping of redirected interrupts

User space requests the tear down of a redirected interrupt using the
APIC ID and the CPU interrupt vector number associated with it.

Traverse the list of redirected interrupt and release the resources of
the IRQ being teared down.

Signed-off-by: Ricardo Neri <ricardo.neri-calderon@linux.intel.com>
---
Changes since v2:
 * Introduced this patch.

Changes since v1:
 * N/A

ricardon commented:
I have updated this pull request to not touch the arch/x86 directory and instead use the existing Linux IRQ infrastructure.

ricardon commented:
@romank-msft @chris-oo would you mind taking a look at these changes when you get a chance? These are the kernel changes that https://github.com/microsoft/openvmm/pull/384 from @balajimc55

chris-oo commented on drivers/hv/mshv_vtl_main.c:
```diff
@@ -2874,6 +2933,10 @@ static int __init mshv_vtl_init(void)
 		goto free_mem;
 	}
 
+	ret = ms_hyperv_init_redirected_intr();
```
I'm wondering if we should only do this on TDX today? Or only TDX + SNP / platforms where we actually expect to do interrupt redirection? How much overhead are we allocating here?

chris-oo replied:
talking with the hypervisor folks, we should gate this functionality to TDX only for now.

ricardon replied:
The function ms_hyperv_init_redirected_intr() is wrapped in #if defined(CONFIG_X86_64) && defined(CONFIG_INTEL_TDX_GUEST). It has a stub for other cases. Do you mean another type of gating?

Regarding the overhead, vmlinux remained the same size. vmlinux.dbg increases by 5KB (0.008%). At runtime, there should be no overhead since the compiler inlines back the code I carved out into the new function do_assert_single_proxy_intr(). During boot, ms_hyperv_init_redirected_intr() creates an irq_domain and a linked list that would stay unused.

I did not test on an SNP platform. The code should work with any x86_64 system, though.

chris-oo replied:
Ah no, I missed that from the PR view. That sounds good to me though.

ricardon replied:
Thanks! @dcui do you have any feedback?

chris-oo reviewed (commented):
mostly looks good, just had a question on when we should enable this functionality

namancse commented:
Changes look good to me. They are well separated into different commits, and the commit messages make it easy to understand the purpose and implementation.

namancse reviewed (commented):
Looks good to me.

ricardon commented:
> Looks good to me.

Thank you @namancse !

hargar19 commented:
@ricardon Please merge this PR when you're ready

ricardon commented:
@hargar19 I don't think I have permissions to merge PRs.
